File: backend/app/models/term_account.rb
# == Schema Information
#
# Table name: term_accounts
#
#  id                :bigint           not null, primary key
#  account_number    :string(100)
#  account_type      :string           not null
#  amount            :decimal(14, 2)   not null
#  balance           :decimal(14, 2)   default(0.0), not null
#  closed_amount     :decimal(14, 2)
#  closed_date       :date
#  interest_rate     :decimal(5, 2)    not null
#  is_active         :boolean          default(TRUE), not null
#  maturity_amount   :decimal(14, 2)   not null
#  maturity_date     :date             not null
#  notes             :text
#  open_date         :date             not null
#  tenure_days       :integer
#  created_at        :datetime         not null
#  updated_at        :datetime         not null
#  parent_account_id :bigint           not null
#  user_id           :bigint           not null
#
# Indexes
#
#  index_term_accounts_on_parent_account_id  (parent_account_id)
#  index_term_accounts_on_user_id            (user_id)
#
# Foreign Keys
#
#  fk_rails_...  (parent_account_id => accounts.id) ON DELETE => restrict
#  fk_rails_...  (user_id => users.id) ON DELETE => cascade
#
class TermAccount < ApplicationRecord
  class Error < StandardError; end

  audited only: [ :balance ]

  has_one :import_record, as: :importable, dependent: :nullify

  belongs_to :user
  belongs_to :parent_account, class_name: "Account"

  enum :account_type, { fd: "fd", ppf: "ppf" }, validate: true

  scope :active, -> { where(is_active: true) }
  scope :inactive, -> { where(is_active: false) }

  validates :amount,          presence: true, numericality: { greater_than: 0 }
  validates :open_date,       presence: true
  validates :interest_rate,   presence: true
  validates :maturity_date,   presence: true
  validates :maturity_amount, presence: true
  validates :tenure_days,     presence: true, if: :fd?

  before_validation :apply_defaults, on: :create

  def closed?
    !is_active
  end

  def deposit!(amount)
    raise Error, "Term account is closed" if closed?
    update!(balance: balance + amount)
  end

  def close!(closed_date:, closed_amount:)
    raise Error, "Term account is already closed" if closed?
    # Stamp the audit row so the Balance History sidebar can label this
    # "Account closed" instead of an anonymous "Balance update".
    self.audit_comment = "close:term_account_#{id}"
    update!(closed_date: closed_date, closed_amount: closed_amount, balance: 0, is_active: false)
  end

  private

  def apply_defaults
    self.account_number ||= if fd?
      Time.current.strftime("FD#%Y%m%d%H%M")
    else
      Time.current.strftime("PPF#%Y%m%d%H%M")
    end

    if maturity_date.nil? && open_date
      self.maturity_date = if fd? && tenure_days
        open_date + tenure_days.days
      elsif ppf?
        open_date >> (15 * 12)
      end
    end

    if fd? && maturity_amount.nil? && amount && interest_rate && tenure_days
      years = tenure_days.to_f / 365
      self.maturity_amount = (amount * (1 + interest_rate / 400.0) ** (4 * years)).round(2)
    end
  end
end

File: backend_python/app/models/term_account.py
import enum
from datetime import date, datetime, timedelta

from app.config import TZ

from sqlalchemy import Boolean, Date, DateTime, Enum, ForeignKey, Integer, Numeric, String, event, func
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models._audit_registry import auditable


class TermAccountType(str, enum.Enum):
    fd = "fd"
    ppf = "ppf"


@auditable("balance")
class TermAccount(Base):
    __tablename__ = "term_accounts"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    parent_account_id: Mapped[int] = mapped_column(
        ForeignKey("accounts.id", ondelete="RESTRICT"), nullable=False
    )
    type: Mapped[TermAccountType] = mapped_column(
        Enum(TermAccountType, name="term_account_type", create_type=False), nullable=False
    )
    account_number: Mapped[str | None] = mapped_column(String(100), nullable=True)
    amount: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    open_date: Mapped[date] = mapped_column(Date, nullable=False)
    tenure_days: Mapped[int | None] = mapped_column(Integer, nullable=True)
    interest_rate: Mapped[float] = mapped_column(Numeric(5, 2), nullable=False)
    maturity_date: Mapped[date] = mapped_column(Date, nullable=False)
    maturity_amount: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    balance: Mapped[float] = mapped_column(
        Numeric(14, 2), nullable=False, default=0, server_default="0"
    )
    closed_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    closed_amount: Mapped[float | None] = mapped_column(Numeric(14, 2), nullable=True)
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    parent_account = relationship("Account", foreign_keys=[parent_account_id])
    user = relationship("User")
    bank = association_proxy("parent_account", "bank")
    transactions = relationship(
        "Transaction",
        primaryjoin="and_(TermAccount.id == foreign(Transaction.linked_account_id), Transaction.linked_account_type == 'term_account')",
        viewonly=True,
    )
    audit_logs = relationship(
        "AuditLog",
        primaryjoin="and_(TermAccount.id == foreign(AuditLog.record_id), AuditLog.table_name == 'term_accounts')",
        viewonly=True,
    )

    def deposit(self, amount: float) -> None:
        if not self.is_active:
            raise ValueError(f"Term account '{self.account_number or self.id}' is closed")
        self.balance = float(self.balance) + amount

    def close(self, closed_date: date, closed_amount: float) -> None:
        if not self.is_active:
            raise ValueError(f"Term account '{self.account_number or self.id}' is already closed")
        self.closed_date = closed_date
        self.closed_amount = closed_amount
        self.balance = 0
        self.is_active = False

    def apply_defaults(self) -> None:
        if not self.account_number:
            if self.type == TermAccountType.fd:
                self.account_number = datetime.now(TZ).strftime("FD#%Y%m%d%H%M")
            elif self.type == TermAccountType.ppf:
                self.account_number = datetime.now(TZ).strftime("PPF#%Y%m%d%H%M")

        if self.maturity_date is None:
            if self.type == TermAccountType.fd and self.tenure_days:
                self.maturity_date = self.open_date + timedelta(days=self.tenure_days)
            elif self.type == TermAccountType.ppf and self.open_date:
                self.maturity_date = self.open_date.replace(year=self.open_date.year + 15)

        if self.type == TermAccountType.fd:
            # Compound interest (quarterly) — only when not provided by caller
            if not self.maturity_amount and self.amount and self.interest_rate and self.tenure_days:
                years = float(self.tenure_days) / 365
                self.maturity_amount = round(
                    float(self.amount) * (1 + float(self.interest_rate) / 400) ** (4 * years), 2
                )


@event.listens_for(TermAccount, "before_insert")
def _set_term_account_defaults(mapper, connection, target: TermAccount) -> None:
    target.apply_defaults()


